src/Extensions/Compatability/Validation/Code.cs

```diff
@@ -0,0 +1,515 @@
+		/// <param name="state">Object's state</param>
+		/// <param name="errorMessage">Error message</param>
+		/// <param name="tagId">Tag Id to log, leave null if no logging is needed</param>
+		[Obsolete(ObsoleteMessage, IsObsoleteError)]
+		public static void Expects<TException>(bool state, string errorMessage, EventId? tagId)
+			where TException : Exception
+		{
+			if (!Validate(state, errorMessage, tagId))
+			{
+				ReportError<TException>(errorMessage);
+			}
+		}
+
+
+		/// <summary>
+		/// Checks that the enumerable argument is not null and doesn't contain any nulls
+		/// </summary>
+		/// <remarks>Be careful to not pass enumerables that can be enumerated only once</remarks>
+		/// <typeparam name="TEnumerable">Type of the enumerable</typeparam>
+		/// <param name="argumentValue">The argument value.</param>
+		/// <param name="argumentName">Name of the argument.</param>
+		/// <param name="tagId">Tag Id to log, leave null if no logging is needed</param>
+		/// <exception cref="ArgumentException">Thrown if any argument  <paramref name="argumentValue"/> element is null.</exception>
+		/// <exception cref="ArgumentNullException">Thrown if the supplied argument <paramref name="argumentValue"/> is null.</exception>
+		[Obsolete(ObsoleteMessage, IsObsoleteError)]
+		public static IEnumerable<TEnumerable> ExpectsAllNotNull<TEnumerable>([ValidatedNotNull] IEnumerable<TEnumerable>? argumentValue, string argumentName, EventId? tagId)
+			where TEnumerable : class
+		{
+			argumentValue = ExpectsArgumentNotNull(argumentValue, argumentName, tagId);
+
+			if (!ValidateAllNotNull(argumentValue, argumentName, tagId))
+			{
+				ReportArgumentError(argumentName, AllErrorMessage);
+			}
+
+			return argumentValue;
+		}
+
+
+		/// <summary>
+		/// Checks the collection and throws an exception if it is null, contains no values, or contains any null values
+		/// </summary>
+		/// <typeparam name="T">Type of the collection</typeparam>
+		/// <param name="argumentValue">The argument value.</param>
+		/// <param name="argumentName">Name of the argument.</param>
+		/// <param name="tagId">Tag Id to log, leave null if no logging is needed.</param>
+		/// <exception cref="ArgumentException">Thrown if <paramref name="argumentValue"/> is empty or if any element is null.</exception>
+		/// <exception cref="ArgumentNullException">Thrown if the supplied argument <paramref name="argumentValue"/> is null.</exception>
+		[Obsolete(ObsoleteMessage, IsObsoleteError)]
+		public static IEnumerable<T> ExpectsNotEmptyAndAllNotNull<T>([ValidatedNotNull] ICollection<T>? argumentValue, string argumentName, EventId? tagId) where T : class
+		{
+			argumentValue = ExpectsArgumentNotNull(argumentValue, argumentName, tagId);
+
+			if (!ValidateNotEmptyAndAllNotNull(argumentValue, argumentName, tagId))
+			{
+				ReportArgumentError(argumentName, HasAnyErrorMessage);
+			}
+
+			return argumentValue;
+		}
+
+
+		/// <summary>
```

Comment:
Typeparam not documented and poorly named

Reply:
explained here https://github.com/microsoft/Omex/pull/153#discussion_r390509059